docs/proposals/kubelet-qos.md

```diff
@@ -0,0 +1,142 @@
+- Pod infrastructure container
+  - OOM_SCORE_ADJ: -999
+- Kubelet, Docker, Kube-Proxy
+  - OOM_SCORE_ADJ: -1000 (won’t be OOM killed)
+  - Hack, in the future, we should place all pods into a separate cgroup, and set a limit on the memory they can consume
+
+Ensuring resource guarantees
+- After being configured suitably, the kernel will handle resource guarantees as specified in QoS Functionality
+
+## Issues and Assumptions
+
+- The QoS proposal assumes that swap memory is disabled. If swap is enabled, then resource guarantees (for pods that specify resource requirements) will not hold. For example, suppose 2 top-tier pods have reached their memory limit. They can start allocating memory on swap space. Eventually, if there isn’t enough swap space, processes in the pods might get killed. TODO: ensure that swap space is disabled on our cluster setups scripts.
+- We cannot set OOM_SCORE_ADJ of a process until it has launched. This could lead to races. For example, suppose that a burstable pod is using 70% of the system’s memory, and another burstable pod is using 30% of the system’s memory. A best-effort pod attempts to launch on the Kubelet. Initially the best-effort pod is using 2% of memory, and has an OOM_SCORE_ADJ of 0. So its OOM_SCORE is lower than the burstable  pod using 70% of system memory. The burstable pod will be evicted by the best-effort pod. Short-term TODO: Consider implementing a restart policy where best-effort pods are immediately evicted if OOM killed, but burstable pods are given a few retries. Long-term TODO: push support for OOM scores in cgroups to the upstream Linux kernel. This is out of scope.
+- We need to deal with the Kubelet’s restart policy when a process gets killed by the kernel. Suppose that we have 30 best-effort pods wanting to use 1GB of memory, with an “always” restart policy, in a system with 8GB of memory. In the current proposal, the system will keep launching best-effort pods, and the kernel will keep killing best-effort pods. In fact, the OOM killer is biased towards killing pods that are consuming more memory, so best-effort pods that are running would be evicted. In the first iteration, we will encourage users not to use the “always” restart policy with best-effort pods, but won’t handle the restart issue in Kubelet. Extension: In the final implementation, we will not allow best-effort pods to restart, and will limit the number of times burstable pods can restart if OOM killed.
+- The OOM killer might kill processes across different pods. Suppose that 10 best-effort pods are running on the Kubelet, and the system runs out of memory. The OOM killer might kill one process from each pod. Short-term TODO: Docker’s philosophy has been not to care about this issue, so for now we will not focus on it. Long-term TODO: Treat a pod as a unit - so that one entire pods is killed, instead of a few processes across different pods.

```

Comment:
We should at least terminate the whole container. Yes, this isn't ideal for containers-as-VMs that run process supervisors, but one of the reasons we have pods is so that users don't need to do that. We can add a knob to control this behavior later, if necessary. Besides, if we don't kill the container, we can't apply the pod-level termination behavior.

By the way, this is a good place to coin the term "out-of-resource killing".

TODO: Develop a comparable policy for CPU starvation.